sql/core/src/main/scala/org/apache/spark/sql/scripting/SqlScriptingExecutionNode.scala

```diff
@@ -649,3 +652,145 @@ class LoopStatementExec(
+    queryResult
+  }
+
+  /**
+   * Loop can be interrupted by LeaveStatementExec
+   */
+  private var interrupted: Boolean = false
+
+  private lazy val treeIterator: Iterator[CompoundStatementExec] =
+    new Iterator[CompoundStatementExec] {
+      override def hasNext: Boolean =
+        !interrupted && cachedQueryResult().length > 0 && currRow < cachedQueryResult().length
+
+      override def next(): CompoundStatementExec = state match {
+        case ForState.VariableDeclaration =>
+          // when there is no for variable, skip var declaration and iterate only the body
+          if (variableName.isEmpty) {
+            state = ForState.Body
+            body.reset()
+            return next()
+          }
+          currVariable = createExpressionFromValue(cachedQueryResult()(currRow))
+          state = ForState.VariableAssignment
+          createDeclareVarExec(variableName.get, currVariable)
+
+        case ForState.VariableAssignment =>
+          state = ForState.Body
+          body.reset()
+          createSetVarExec(variableName.get, currVariable)
+
+        case ForState.Body =>
+          val retStmt = body.getTreeIterator.next()
+
+          // Handle LEAVE or ITERATE statement if it has been encountered.
+          retStmt match {
+            case leaveStatementExec: LeaveStatementExec if !leaveStatementExec.hasBeenMatched =>
+              if (label.contains(leaveStatementExec.label)) {
+                leaveStatementExec.hasBeenMatched = true
+              }
+              interrupted = true
+              return retStmt
+            case iterStatementExec: IterateStatementExec if !iterStatementExec.hasBeenMatched =>
+              if (label.contains(iterStatementExec.label)) {
+                iterStatementExec.hasBeenMatched = true
+              }
+              currRow += 1
+              state = ForState.VariableDeclaration
+              return retStmt
+            case _ =>
+          }
+
+          if (!body.getTreeIterator.hasNext) {
+            currRow += 1
+            state = ForState.VariableDeclaration
+          }
+          retStmt
+      }
+    }
+
+  /**
+   * Creates a Catalyst expression from Scala value.
+   */
+  private def createExpressionFromValue(value: Any): Expression = value match {
+    case m: Map[_, _] =>
+      // arguments of CreateMap are in the format: (key1, val1, key2, val2, ...)
+      val mapArgs = m.keys.toSeq.flatMap { key =>
+        Seq(createExpressionFromValue(key), createExpressionFromValue(m(key)))
+      }
+      CreateMap(mapArgs, false)
+    case s: GenericRowWithSchema =>
+    // struct types match this case
+    // arguments of CreateNamedStruct are in the format: (name1, val1, name2, val2, ...)
+    val namedStructArgs = s.schema.names.toSeq.flatMap { colName =>
+        val valueExpression = createExpressionFromValue(s.getAs(colName))
+        Seq(Literal(colName), valueExpression)
+      }
+      CreateNamedStruct(namedStructArgs)
+    case _ => Literal(value)
```

Comment:
for my knowledge, can you explain what does the case with the `Map` means exactly, i.e. when will this happen?
also, how did we check that this is the complete list of the relevant cases?